ops/pebble.py

```diff
@@ -308,29 +317,97 @@ def __repr__(self):
                 ).format(self=self)
 
 
-class API:
+class Setup:
+    """Represents a Pebble setup layer (or flattened setup).
+
+    This is not documented, but captured in code here:
+    https://github.com/canonical/pebble/blob/master/internal/setup/setup.go
+    """
+
+    def __init__(self, raw: Union[str, Dict] = None):
+        if isinstance(raw, str):
+            d = yaml.safe_load(raw) or {}
+        else:
+            d = raw or {}
+        self.summary = d.get('summary', '')
+        self.description = d.get('description', '')
+        self.services = {name: SetupService(name, service)
+                         for name, service in d.get('services', {}).items()}
+
+    def to_yaml(self) -> str:
+        """Convert this setup to its YAML representation."""
+        return yaml.dump(self.to_dict(), Dumper=_safe_dumper)
+
+    def to_dict(self) -> Dict:
+        """Convert this setup to a dict representation."""
+        fields = [
+            ('summary', self.summary),
+            ('description', self.description),
+            ('services', {name: service.to_dict() for name, service in self.services.items()})
+        ]
+        return {name: value for name, value in fields if value}
+
+    def __repr__(self) -> str:
+        return 'Setup({!r})'.format(self.to_dict())
+
+    __str__ = to_yaml
+
+
+class SetupService:
+    """Represents a service description in a Pebble setup layer."""
+
+    def __init__(self, name: str, raw: Dict = None):
+        self.name = name
+        raw = raw or {}
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        self.default = raw.get('default', '')
+        self.override = raw.get('override', '')
+        self.command = raw.get('command', '')
+        self.after = raw.get('after', [])
+        self.before = raw.get('before', [])
+        self.requires = raw.get('requires', [])
+        self.environment = raw.get('environment') or {}
```

Comment:
What's the borrowing semantics we intend for this type?  Do we care? That is, if people provide a raw value in, are we a) copying it; or b) taking it over and potentially allowing further code to mutate it?

Reply:
Good call. I've changed these (also the lists) to make copies and added a test for that.